man/ngettext.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.24.
.TH NGETTEXT "1" "June 2001" "GNU gettext 0.11" GNU
.SH NAME
ngettext \- translate message and choose plural form
.SH SYNOPSIS
.B ngettext
[\fIOPTION\fR] [\fITEXTDOMAIN\fR] \fIMSGID MSGID-PLURAL COUNT\fR
.SH DESCRIPTION
.\" Add any additional description here
The \fBngettext\fP program translates a natural language message into the
user's language, by looking up the translation in a message catalog, and
chooses the appropriate plural form, which depends on the number \fICOUNT\fP
and the language of the message catalog where the translation was found.
.PP
Display native language translation of a textual message whose grammatical
form depends on a number.
.TP
\fB\-d\fR, \fB\-\-domain\fR=\fITEXTDOMAIN\fR
retrieve translated message from TEXTDOMAIN
.TP
\fB\-e\fR
enable expansion of some escape sequences
.TP
\fB\-E\fR
(ignored for compatibility)
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
display version information and exit
.TP
[TEXTDOMAIN]
retrieve translated message from TEXTDOMAIN
.TP
MSGID MSGID-PLURAL
translate MSGID (singular) / MSGID-PLURAL (plural)
.TP
COUNT
choose singular/plural form based on this value
.PP
If the TEXTDOMAIN parameter is not given, the domain is determined from the
environment variable TEXTDOMAIN.  If the message catalog is not found in the
regular directory, another location can be specified with the environment
variable TEXTDOMAINDIR.
Standard search directory: /home/haible/gnu/arch/linuxlibc6/share/locale
.SH AUTHOR
Written by Ulrich Drepper.
.SH "REPORTING BUGS"
Report bugs to <bug-gnu-utils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 1995-1997, 2000, 2001 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B ngettext
is maintained as a Texinfo manual.  If the
.B info
and
.B ngettext
programs are properly installed at your site, the command
.IP
.B info ngettext
.PP
should give you access to the complete manual.


NEWS
Version 1.1 - October 2026

# Programming languages support:
  * C++, Rust:
    - The --keyword option now supports "qualified names" of the form
      'namespace::function', that specify a function in a namespace or
      possibly even a function in a nested namespace.
  * Scheme:
    - A new example 'hello-scheme' has been added.

Version 1.0 - January 2026

# Improvements for maintainers and distributors:
  * In a po/ directory, the PO files are now exactly those that the
    translators submitted or committed in version control, or a
    translation project's daemon committed on behalf of the translators.
    They are no longer regularly updated with respect to the POT file
    in the same directory.

    The advantage for maintainers is that the maintainer may commit the
    PO files in version control, without getting
      - lots of modified files shown by "git status",
      - frequent merge conflicts when merging between branches,
      - a voluminous version control history.

    The advantage for distributors is that the role of files in a
    release tarball are clearer: The PO files are source code, whereas
    the POT file and the *.gmo files are generated files.

    ATTENTION translators!
    Translators who work directly on a package's source code (without
    going through a translation project) now need to run "msginit"
    before starting work on a PO file.

  * A new program 'po-fetch' is provided, that fetches the translated
    PO files from a translation project's site on the internet, and
    updates the LINGUAS file accordingly.

  * In a po/ directory, a new script 'fetch-po' is now added by 'gettextize'.
    It provides the standard interface for fetching the translated PO files.
    It typically either invokes the 'po-fetch' program or does nothing.

# Improvements for translators:
  * msginit:
    - When the PO file already exists, 'msginit' now updates it w.r.t. the
      POT file, like 'msgmerge' would do. Previously, 'msginit' failed with
      an error message in this situation.

  * Pretranslation:
    - Two new programs, 'msgpre' and 'spit', are provided, that implement
      machine translation through a locally installed Large Language Model
      (LLM).  'msgpre' applies to an entire PO file, 'spit' to a single
      message.
    - The documentation has a new chapter "Pretranslation".

# Improvements for maintainers:
  * xgettext:
    - The refactoring suggestion when a translatable string contains an URL
      or email address can now be inhibited through a command-line option
      '--no-check=url' or '--no-check=email', or through a comment in the
      source code of the form
          /* xgettext: no-url-check */
      or
          /* xgettext: no-email-check */

# Programming languages support:
  * OCaml:
    - xgettext now supports OCaml.
    - 'msgfmt -c' now verifies the syntax of translations of OCaml format
      strings.
    - A new example 'hello-ocaml' has been added.
  * Rust:
    - xgettext now recognizes 'gettextrs::gettext' invocations, like 'gettext'
      invocations.

# libgettextpo library:
  * The function 'po_message_get_format' now supports distinguishing whether
    a negative format string mark, such as 'no-c-format', is set or not.
  * The new functions
      po_message_has_workflow_flag
      po_message_set_workflow_flag
      po_message_workflow_flags_iterator, po_flag_next, po_flag_iterator_free
    can be used to manipulate or inspect the workflow flags of a message.
  * The new functions
      po_message_has_sticky_flag
      po_message_set_sticky_flag
      po_message_sticky_flags_iterator, po_flag_next, po_flag_iterator_free
    can be used to manipulate or inspect the sticky flags of a messsage.

# Emacs PO mode:
  Restore syntax highlighting in Emacs version 30 or newer.

Version 0.26 - July 2025

# Programming languages support:
  * JavaScript:
    - xgettext now parses regular expressions with character classes
      correctly.
  * C, C++, Python, JavaScript, EmacsLisp, librep, Go, Ruby, awk, D, Tcl,
    Perl, PHP:
    - xgettext's heuristic recognition of format strings has been improved:
      strings like "100% complete" (with a space flag in a format directive)
      are no longer flagged as format strings by default, unless they occur
      in a context that requires a format string.  You can override this
      heuristic by using a comment of the form /* xgettext: c-format */.
  * Shell:
    - The documentation now mentions two other approaches for
      internationalizing messages with parameters in shell scripts.
    - xgettext now recognizes format strings in the 'printf' command syntax.
      They are marked as 'sh-printf-format' in POT and PO files.
    - Two new programs 'printf_gettext' and 'printf_ngettext' are provided,
      that do formatted output with a localized format string in a more
      efficient way (without spawning a subshell).
    - xgettext now recognizes the \c, \u, and \U escape sequences in dollar-
      single-quoted strings $'...'.

# Improvements for maintainers:
  * xgettext:
    - When extracting a message with plural that is some format string,
      xgettext now verifies that the msgid and msgid_plural are compatible
      as format strings.  For most format string types, this still allows
      omitting from msgid a placeholder that is used in msgid_plural.  But
      when a placeholder is used in both msgid and msgid_plural, its type
      must be the same in both.
    - xgettext now suggests a refactoring when a translatable string
      contains an URL or email address.

# Improvements for translators:
  * msggrep:
    - msggrep accepts two new options -W/--workflow-flags and -S/--sticky-flags
      that allow to select only messages that have a specified flag.

# Bug fixes:
  - The AM_GNU_GETTEXT macro now rejects the dysfunctional gettext() function
    in libc of Solaris 11.[0-3], Solaris OpenIndiana, and Solaris OmniOS.

src/ngettext.c
/* ngettext - retrieve plural form string from message catalog and print it.
   Copyright (C) 1995-1997, 2000, 2001 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <errno.h>

#include "error.h"
#include "basename.h"
#include "xmalloc.h"
#include "exit.h"

#include "gettext.h"

#define _(str) gettext (str)

/* If nonzero expand escape sequences in strings before looking in the
   message catalog.  */
int do_expand;

/* Name the program is called with.  */
const char *program_name;

/* Long options.  */
static const struct option long_options[] =
{
  { "domain", required_argument, NULL, 'd' },
  { "help", no_argument, NULL, 'h' },
  { "version", no_argument, NULL, 'V' },
  { NULL, 0, NULL, 0 }
};

/* Prototypes for local functions.  Needed to ensure compiler checking of
   function argument counts despite of K&R C function definition syntax.  */
static void usage PARAMS ((int status))
#if defined __GNUC__ && ((__GNUC__ == 2 && __GNUC_MINOR__ >= 5) || __GNUC__ > 2)
     __attribute__ ((noreturn))
#endif
;
static const char *expand_escape PARAMS ((const char *str));

int
main (argc, argv)
     int argc;
     char *argv[];
{
  int optchar;
  const char *msgid;
  const char *msgid_plural;
  const char *count;
  unsigned long n;

  /* Default values for command line options.  */
  int do_help = 0;
  int do_version = 0;
  const char *domain = getenv ("TEXTDOMAIN");
  const char *domaindir = getenv ("TEXTDOMAINDIR");
  do_expand = 0;

  /* Set program name for message texts.  */
  program_name = argv[0];
  if (strncmp (program_name, "lt-", 3) == 0)
    program_name += 3;

#ifdef HAVE_SETLOCALE
  /* Set locale via LC_ALL.  */
  setlocale (LC_ALL, "");
#endif

  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  /* Parse command line options.  */
  while ((optchar = getopt_long (argc, argv, "+d:eEhV", long_options, NULL))
	 != EOF)
    switch (optchar)
    {
    case '\0':		/* Long option.  */
      break;
    case 'd':
      domain = optarg;
      break;
    case 'e':
      do_expand = 1;
      break;
    case 'E':
      /* Ignore.  Just for compatibility.  */
      break;
    case 'h':
      do_help = 1;
      break;
    case 'V':
      do_version = 1;
      break;
    default:
      usage (EXIT_FAILURE);
    }

  /* Version information is requested.  */
  if (do_version)
    {
      printf ("%s (GNU %s) %s\n", basename (program_name), PACKAGE, VERSION);
      /* xgettext: no-wrap */
      printf (_("Copyright (C) %s Free Software Foundation, Inc.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
"),
	      "1995-1997, 2000, 2001");
      printf (_("Written by %s.\n"), "Ulrich Drepper");
      exit (EXIT_SUCCESS);
    }

  /* Help is requested.  */
  if (do_help)
    usage (EXIT_SUCCESS);

  /* More optional command line options.  */
  switch (argc - optind)
    {
    default:
      error (EXIT_FAILURE, 0, _("too many arguments"));

    case 4:
      domain = argv[optind++];
      /* FALLTHROUGH */

    case 3:
      break;

    case 2:
    case 1:
    case 0:
      error (EXIT_FAILURE, 0, _("missing arguments"));
    }

  /* Now the mandatory command line options.  */
  msgid = argv[optind++];
  msgid_plural = argv[optind++];
  count = argv[optind++];

  if (optind != argc)
    abort ();

  {
    char *endp;
    unsigned long tmp_val;

    errno = 0;
    tmp_val = strtoul (count, &endp, 10);
    if (errno == 0 && count[0] != '\0' && endp[0] == '\0')
      n = tmp_val;
    else
      /* When COUNT is not valid, use plural.  */
      n = 99;
  }

  /* Expand escape sequences if enabled.  */
  if (do_expand)
    {
      msgid = expand_escape (msgid);
      msgid_plural = expand_escape (msgid_plural);
    }

  /* If no domain name is given we don't translate, and we use English
     plural form handling.  */
  if (domain == NULL || domain[0] == '\0')
    fputs (n == 1 ? msgid : msgid_plural, stdout);
  else
    {
      /* Bind domain to appropriate directory.  */
      if (domaindir != NULL && domaindir[0] != '\0')
	bindtextdomain (domain, domaindir);

      /* Write out the result.  */
      fputs (dngettext (domain, msgid, msgid_plural, n), stdout);
    }

  exit (EXIT_SUCCESS);
}


/* Display usage information and exit.  */
static void
usage (status)
     int status;
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      /* xgettext: no-wrap */
      printf (_("\
Usage: %s [OPTION] [TEXTDOMAIN] MSGID MSGID-PLURAL COUNT\n\
"), program_name);
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
Display native language translation of a textual message whose grammatical\n\
form depends on a number.\n"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
  -d, --domain=TEXTDOMAIN   retrieve translated message from TEXTDOMAIN\n\
  -e                        enable expansion of some escape sequences\n\
  -E                        (ignored for compatibility)\n\
  -h, --help                display this help and exit\n\
  -V, --version             display version information and exit\n\
  [TEXTDOMAIN]              retrieve translated message from TEXTDOMAIN\n\
  MSGID MSGID-PLURAL        translate MSGID (singular) / MSGID-PLURAL (plural)\n\
  COUNT                     choose singular/plural form based on this value\n"));
      printf ("\n");
      /* xgettext: no-wrap */
      printf (_("\
If the TEXTDOMAIN parameter is not given, the domain is determined from the\n\
environment variable TEXTDOMAIN.  If the message catalog is not found in the\n\
regular directory, another location can be specified with the environment\n\
variable TEXTDOMAINDIR.\n\
Standard search directory: %s\n"),
	      getenv ("IN_HELP2MAN") == NULL ? LOCALEDIR : "@localedir@");
      printf ("\n");
      fputs (_("Report bugs to <bug-gnu-gettext@gnu.org>.\n"), stdout);
    }

  exit (status);
}


/* Expand some escape sequences found in the argument string.  */
static const char *
expand_escape (str)
     const char *str;
{
  char *retval, *rp;
  const char *cp = str;

  do
    {
      while (cp[0] != '\0' && cp[0] != '\\')
	++cp;
    }
  while (cp[0] != '\0' && cp[1] != '\0'
	 && strchr ("bfnrt\\01234567", cp[1]) == NULL);

  if (cp[0] == '\0')
    return str;

  retval = (char *) xmalloc (strlen (str));

  rp = retval + (cp - str);
  memcpy (retval, str, cp - str);

  do
    {
      switch (*++cp)
	{
	case 'b':		/* backspace */
	  *rp++ = '\b';
	  ++cp;
	  break;
	case 'f':		/* form feed */
	  *rp++ = '\f';
	  ++cp;
	  break;
	case 'n':		/* new line */
	  *rp++ = '\n';
	  ++cp;
	  break;
	case 'r':		/* carriage return */
	  *rp++ = '\r';
	  ++cp;
	  break;
	case 't':		/* horizontal tab */
	  *rp++ = '\t';
	  ++cp;
	  break;
	case '\\':
	  *rp = '\\';
	  ++cp;
	  break;
	case '0': case '1': case '2': case '3':
	case '4': case '5': case '6': case '7':
	  {
	    int ch = *cp++ - '0';

	    if (*cp >= '0' && *cp <= '7')
	      {
		ch *= 8;
		ch += *cp++ - '0';

		if (*cp >= '0' && *cp <= '7')
		  {
		    ch *= 8;
		    ch += *cp++ - '0';
		  }
	      }
	    *rp = ch;
	  }
	  break;
	default:
	  *rp = '\\';
	  break;
	}

      while (cp[0] != '\0' && cp[0] != '\\')
	*rp++ = *cp++;
    }
  while (cp[0] != '\0');

  /* Terminate string.  */
  *rp = '\0';

  return (const char *) retval;
}
